base/reflection.jl

```diff
@@ -672,21 +671,31 @@ end
 Determine whether type `T` has exactly one possible instance; for example, a
 struct type with no fields.
 """
-issingletontype(@nospecialize(t)) = (@_pure_meta; isa(t, DataType) && isdefined(t, :instance))
+issingletontype(@nospecialize(t)) = (@_total_meta; isa(t, DataType) && isdefined(t, :instance))
 
 """
     typeintersect(T, S)
 
 Compute a type that contains the intersection of `T` and `S`. Usually this will be the
 smallest such type or one close to it.
 """
-function typeintersect(@nospecialize(a), @nospecialize(b))
-    @_pure_meta
+function typeintersect end
+@eval function typeintersect(@nospecialize(a), @nospecialize(b))
     (!isvarargtype(a) && !isvarargtype(b)) || throw(ArgumentError("invalid typeintersect with Vararg"))
-    return ccall(:jl_type_intersection, Any, (Any, Any), a, b)
+    # equivalent to: @assume_effects :total @ccall jl_type_intersection(a::Any, b::Any)::Any
+    return $(Expr(:foreigncall, QuoteNode(:jl_type_intersection),
```

Comment:
How feasible would it be to define the ccall macro earlier or maybe have a limited version of it. I was fine with one or two of those, but if this expands, it's a bit of a nightmare.

Reply:
Might even be worth making it a non-proper scheme macro (not handled by macroexpand), since it can be rather complex to unpack the various possible arguments, but therefore also worthwhile to have very early in bootstrapping

Reply:
I spent a bit time to rearrange utilities of essentials.jl/reflection.jl/expr.jl, but didn't succeed to bootstrap. The most difficult part might be that operators.jl isn't available at a time of using `@ccall` in reflection.jl.

For these specific utilities, I think we can use `@_total_but_may_throw_meta` instead and it won't be that bad.